### ISSUE TITLE
Client protocol support "holes" that disable TLS 1.3 implicitly lead to handshake failure with TLS 1.3 servers

### ISSUE THREAD

**Issue by vdukhovni** [openssl/openssl#6964]
When I configure a client with a legacy TLS 1.2 protocol exclusion, e.g. by setting `SSL_OP_NO_TLSv1_2`(rather than by using the new min/max version interface), as a result of the new TLS 1.3 protocol support, configurations that previously negotiated "up to" TLS 1.1, now fail when communicating with a TLS 1.3 server:
```
$ posttls-finger -c -p '!TLSv1.2' "[127.0.0.1]"
posttls-finger: SSL_connect error to 127.0.0.1[127.0.0.1]:25: -1
posttls-finger: warning: TLS library problem: error:1425F175:SSL routines:ssl_choose_client_version:inappropriate fallback:../openssl/ssl/statem/statem_lib.c:1939:
```
If I then also explicitly disable `TLSv1.3` the connection succeeds:
```
$ posttls-finger -c -lmay -Lsummary -p '!TLSv1.2:!TLSv1.3' "[127.0.0.1]"
posttls-finger: Anonymous TLS connection established to 127.0.0.1[127.0.0.1]:25: TLSv1.1 with cipher AECDH-AES256-SHA (256/256 bits)
```
The client should figure out that TLS 1.3 has been (implicitly due to the hole) disabled, and not trigger fallback downgrade detection.

**Comment by davidben** [comment]
This sounds familiar. We end up "compiling" the `SSL_OP_NO_*` functions, combined with `SSL_set_{min,max}_proto_version`, into final min/max versions on both client and server. 
https://boringssl-review.googlesource.com/c/boringssl/+/8513

There's otherwise a lot of weirdness where some parts of the code act on a contiguous range and some parts allow holes and you might get into odd situations like the ClientHello skipping setting up key_shares but then accepting a TLS 1.3 ServerHello and crash because some invariant is off.

**Comment by vdukhovni** [comment]
@davidben I'm also in favour of a consistent contiguous range on clients, that should yield further simplification.  I don't know why servers need a contiguous range, they just need to pick a version that the client supports, and I don't know why "holes" are bad on the server. Is there a corner case I'm missing?  Or new TLS 1.3-specific issues?

**Comment by davidben** [comment]
I primarily did it out of simplicity. It's simplest to say client and server have the same behavior and saves thinking about and testing extra cases that don't actually serve any purpose as far as I can tell.

But some of the ad-hoc anti-downgrade mechanisms also assume this. Consider a {1.0, 1.2} server that sees a 1.0+FALLBACK_SCSV ClientHello. Per the letter of the spec, it should abort. But if the client's true max version is TLS 1.1, this isn't quite right. TLS 1.0 is the correct version. I expect the TLS 1.3 mechanism may have similar issues, if not now, in the future when TLS 1.4 exists and one makes a {1.2, 1.4} server. It's similar to why we couldn't do the server-random signal in draft versions. Drafts are effectively non-contiguous versions.

I believe the broken assumption is this: if we allow non-contiguous ranges for *either* party, it is no longer the case that "client supports something newer than X" and "server supports something newer than X" implies "negotiating X indicates a downgrade attack".

**Comment by vdukhovni** [comment]
@davidben Of course this would also fail if the server did support 1.1 and the client had to do a fallback. Presumably, the fallback is not actually caused by the server not supporting 1.1, since would have then negotiated 1.0 on the initial handshake (where there's no client FALLBACK_SCSV).  So in addition to the server protocol hole, there has to be a middle-box on path that breaks with the ClientHello carrying the client's max supported version, and this should fail as a downgrade.

While making the logic symmetric could be simpler or more clear, it would also needlessly lead to servers using lower protocol versions than available, and can be seen as a regression: someone who disabled TLS 1.1, while leaving 1.0 and 1.2 enabled would now lose.  So I don't think we can do that and claim compatibility between 1.1.0 and 1.1.1.

**Comment by davidben** [comment]
If the server did support 1.1 and the client had to do a fallback, you would want FALLBACK_SCSV to trigger because they should have negotiated TLS 1.1. If the server, however, did *not* enable TLS 1.0 

You're right that this complexity in FALLBACK_SCSV is not needed if you assume the server just negotiates things correctly. I think FALLBACK_SCSV was designed assuming that even new servers may have unknown version negotiation bugs hidden in there somewhere. One could have simply defined FALLBACK_SCSV to be "if you see it, first make sure you've first your version negotiation bugs and then unconditionally abort". Instead it is defined to be a signal from the client: "I support a version of TLS higher than this one; if you do too, clearly someone went wrong".

This is, however, not the case with the TLS 1.3 downgrade signal. It is not conditioned on any sort of remedial fallback, yet it relies on the same conditions. Suppose a server enabled {1.2, 1.4}. Should it send the downgrade signal when negotiating TLS 1.2? If yes, {1.2, 1.3} clients will fail to connect to it. If no, {1.2, 1.3, 1.4} clients will not have downgrade protection.

You're probably stuck implementing the latter if you continue to support gaps, but I will note the FALLBACK_SCSV implementation in OpenSSL right now implements the former.

> So I don't think we can do that and claim compatibility between 1.1.0 and 1.1.1.

I'll leave that to you all. Compatibility isn't really a binary thing. It's a probability. Especially with all the differences, a moderately strong claim of compatibility precludes enabling TLS 1.3 by default. A truly strong claim of compatibility precludes fixing any bugs at all. Both of these are undesirable, so one must balance the benefit against the breakage risk. For BoringSSL, the answer was clear, but we are also more okay with breakage risk than OpenSSL, so the answer may well differ there.

(Though I do think the breakage risk here is much lower than the breakage risk of enabling TLS 1.3 by default. I'll note even BoringSSL hasn't done that yet. We intend to, but not until we've had more experience enabling it in a few more consumers first.)

**Comment by davidben** [comment]
> Suppose a server enabled {1.2, 1.4}

It occurs to me this isn't just the case with a hypothetical TLS 1.4. TLS 1.3 also retroactively adds a downgrade signal from 1.2 down to 1.1. It's merely a SHOULD rather than a MUST for 1.2 implementations, but suppose one decides to implement it. Now decrement all the versions in my example:

Suppose a server enabled {1.1, 1.3}. Should it send the 1.2->1.1 downgrade signal when negotiating TLS 1.1? If yes, {1.1, 1.2} clients will fail to connect to it. If no, {1.1, 1.2, 1.3} clients will not have downgrade protection. They would if you send the 1.3->1.2 signal, but I'll note that's solidly not what the specification says.

>    If negotiating TLS 1.2, TLS 1.3 servers MUST set the last 8 bytes of
>   their Random value to the bytes:
>
>     44 4F 57 4E 47 52 44 01
>
>   If negotiating TLS 1.1 or below, TLS 1.3 servers MUST, and TLS 1.2
>   servers SHOULD, set the last 8 bytes of their ServerHello.Random
>   value to the bytes:
>
>     44 4F 57 4E 47 52 44 00

There's no reading of that text that allows a {1.1, 1.3} server to send the first value when negotiating TLS 1.1.

**Comment by vdukhovni** [comment]
> @davidben There's no reading of that text that allows a {1.1, 1.3} server to send the first value when negotiating TLS 1.1.

Yes, sadly the specification does not anticipate servers having to deal with "holes" that force them to downgrade below a client version that falls into the hole. In similar cases, I tend to implement what is "right" rather than what the spec says, provided that the outcome is interoperable, but here that does not appear to be an option, we can't expect the client to check for a magic value that is not the one corresponding to the negotiated protocol.

Now it is not entirely fair to say that the client "loses downgrade protection", it is still covered by the fallback SCSV, what's lost is a signature by the server over the fallback mechanism payload, so that if some MiTM can downgrade the protocol and strip the fallback SCSV, and yet somehow still manufacture a valid "finished" message (i.e. a major downgrade weakness in TLS 1.0, 1.1 or 1.2), then we don't have additional protection from TLS 1.3.

So I am not sure whether the rather remote downgrade risk warrants incompatible behaviour for the few servers with holes in their protocol list.  Note that by inferring a contiguous range we'd lose the upper half of the protocol list, thereby ensuring a downgrade!  So all in all, I'm inclined to leave hold support on, if you don't see a major flaw in my thinking...  Thanks for the feedback, it is very much appreciated!

**Comment by davidben** [comment]
A minor correction and two points I forgot about:

* There *is* a non-trivial weakness to the 1.2 Finished check. It's overly sensitive to the key exchange, which exacerbated some recent attacks. This isn't fatal as one could say clients and servers must disable all weak key exchanges, and indeed that has been the answer thus far. But it's something the signal was meant to harden. (Otherwise I agree. I should have said the client loses the additional downgrade protection.)

* It's not quiiiite true that the downgrade is additive to the Finished check if clients implement False Start, which is common in the browser world. This also isn't fatal. The impact is some client data is encrypted over whatever one's False Start requirements are, which one hopes are okay! We require ECDHE+AEAD+1.2 in Chrome, our strongest settings at the time. Now 1.3 is strongest, but most servers are 1.2 so losing False Start is not attractive. Instead we say that we hit the strongest settings for 1.2 and will never wrongly False Start at 1.2 with a 1.3 server that implements downgrade signal.

* My feeling is folks with holes most likely didn't mean for them. In particular, consider a server that wanted a max version of 1.1 prior to SSL_CTX_set_max_proto_version. It would set SSL_NO_TLSv1_2 and not SSL_OP_NO_TLSv1_3 because that didn't exist yet. It probably did not want to suddenly enable TLS 1.3 on upgrade. Of course a server that wanted max 1.2 would similarly be confused with no hope of repair, but so it goes.

But otherwise your reasoning makes sense. I personally think keeping things contiguous still makes sense on balance, but if you are unconvinced I will get out of your hair unless I come up with something new to add. :-)

If you do support holes though, you should probably revisit both the FALLBACK_SCSV and the 1.3 server-random signal and make sure they make sense in that light, since the specification text may not quite have considered that case.

**Comment by kaduk** [comment]
At least some parts of the specification consider the case of holes, though probably not all, yes:

```
  [...] Note that this mechanism makes it possible to
   negotiate a version prior to TLS 1.2 if one side supports a sparse
   range.  Implementations of TLS 1.3 which choose to support prior
   versions of TLS SHOULD support TLS 1.2. [...]
```

**Comment by mattcaswell** [comment]
Fix in #7013.
